dipy/reconst/weights_method.py

```diff
@@ -0,0 +1,276 @@
+            design_matrix, leverages, C, cutoff
+        )
+        robust = np.logical_not(cond)
+
+        w[~robust] = 0.0
+        w[robust] = 1.0
+
+    if idx == total_idx:  # WLS without outliers
+        robust = last_robust
+        w[~robust] = 0.0
+        w[robust] = pred_sig[robust == 1]**2
+
+    w[np.isinf(w)] = 0
+    w[np.isnan(w)] = 0
+
+    return w, robust
+
+
+def weights_method_nlls_m_est(data, pred_sig, design_matrix, leverages,
+                              idx, total_idx, last_robust,
+                              m_est="gm", cutoff=3,
+                              outlier_condition_func=simple_cutoff):
+    """Calculate M-estimator weights for NLLS model.
+
+    Parameters
+    ----------
+    data : ndarray
+        The measured signal.
+    pred_sig : ndarray
+        The predicted signal, given a previous fit.
+        Has the same shape as data.
+    design_matrix : ndarray (g, ...)
+        Design matrix holding the covariants used to solve for the
+        regression coefficients.
+    leverages : ndarray
+        The leverages (diagonal of the 'hat matrix') of the fit.
+    idx : int
+        The current iteration number.
+    total_idx : int
+        The total number of iterations.
+    last_robust : ndarray
+        True for inlier indices and False for outlier indices. Must have the
+        same shape as data.
+    m_est : str, optional.
+        M-estimator weighting scheme to use. Currently,
+        'gm' (Geman-McClure) and 'cauchy' are provided.
+    cutoff : float, optional
+        Cut-off value for defining outliers based on fitting residuals.
+        Will be passed to the outlier_condition_func.
+        Typical example: |residuals| > cut_off x standard_deviation
+    outlier_condition_func : callable, optional
+        A function with args and returns as follows:
+        is_an_outlier =
+          outlier_condition_func(residuals, log_residuals, pred_sig,
+                                 design_matrix, leverages, C, cutoff)
+
+    Notes
+    -----
+    Robust weights are calculated specifically for the NLLS problem.
+    On the last iteration, NLLS is performed without outliers.
+
+    """
+    # check if M-estimator is valid (defined in this function)
+    if m_est not in ["gm", "cauchy"]:
+        raise ValueError("unknown M-estimator choice")
+
+    # min 3 iters: (1) NLLS (2) NLLS with M-weights (3) clean NLLS
+    if total_idx < 3:
+        raise ValueError("require 3+ iterations")
+
+    p, N = design_matrix.shape[-1], data.shape[-1]
+    if N <= p:
+        raise ValueError("Fewer data points than parameters.")
+    factor = 1.4826 * np.sqrt(N / (N - p))
+
+    # handle potential zeros
+    pred_sig[pred_sig < MIN_POSITIVE_SIGNAL] = MIN_POSITIVE_SIGNAL
+
+    # calculate quantities needed for C and w
+    log_pred_sig = np.log(pred_sig)
+    residuals = data - pred_sig
+    log_data = np.log(data)
+    log_residuals = log_data - log_pred_sig
+    z = pred_sig * log_residuals
```

Comment:
Just flagging that ruff has identified this as a variable that is not being used in the code. Sorry for not noticing this earlier, but seems like that might point to something that doesn't make sense?

Reply:
Not sure why ruff say that because it seems to be used the line just after. Can it be confuse by the one letter variable ?

Reply:
Wait, which line? I don't see `z` being used on line 251. It is used on line 150 in similar context, so maybe it was erroneously copy-pasted here?

Reply:
ok I was confused with the line 148-151. You are right, we need a clarification here.

Reply:
There is a function where this is no longer used, it seems. It's used in the WLS weight function, but not in the NLLS weight function. It was copy-pasted _on purpose_ but left in by accident.